source/common/rate_limiter_impl.cc

```diff
@@ -156,22 +156,27 @@ DelegatingRateLimiterImpl::DelegatingRateLimiterImpl(
       random_distribution_generator_(std::move(random_distribution_generator)) {}
 
 bool DelegatingRateLimiterImpl::tryAcquireOne() {
-  if (distributed_start_ == absl::nullopt) {
-    if (rate_limiter_->tryAcquireOne()) {
-      distributed_start_ = timeSource().monotonicTime() + random_distribution_generator_();
-    }
+  const Envoy::MonotonicTime now = timeSource().monotonicTime();
+  sanity_check_pending_release_ = false;
+  if (rate_limiter_->tryAcquireOne()) {
+    const Envoy::MonotonicTime adjusted = now + random_distribution_generator_();
+    distributed_timings_.insert(
```

Comment:
Could you add a comment about the invariant we maintain on `distributed_timings_`? It looks like it's simply ascending order. Also, is it easy to add unit tests that insert multiple distribution values other than 1ns in sorted and nonsorted order to make sure each acquisition always happens at the expected nanosecond?

Also, is `distributed_timings_` ever anticipated to become large? If so, would a tree be better than a vector?

Reply:
> Also, is `distributed_timings_` ever anticipated to become large? If so, would a tree be better than a vector?

Yes I think you are right, but I feel it would become more prudent to optimise this if we find use cases that need to support specifying large ranges of values to distribute timings over. In that case, this could list could grow large. Then we should optimise, but for now I'm happy with correctness, and I'd rather add a documentation note recommending to not configure large values at this point in time.

(For the known use case where people add white noise for desynchronisation purposes, I wouldn't expect this to be a factor, as the distribution range should usually be smaller then the request-release interval).

Reply:
> Also, is it easy to add unit tests that insert multiple distribution values other than 1ns in sorted and nonsorted order to make sure each acquisition always happens at the expected nanosecond?

Done.